doc/source/example-lm.rst

```diff
@@ -0,0 +1,298 @@
+          print("Ray worker at {url} rank {rank}".format(url=url, rank=world_rank))
+          self.url = url
+          self.world_rank = world_rank
+          args.distributed_rank = world_rank
+          args.distributed_init_method = url
+
+          # Add a hook to the original save_checkpoint function to check whether
+          # or not there is new computational resources available. If so, raise
+          # an exception to restart the training process and make use of the new
+          # resources.
+          if args.cpu:
+              original_n_cpus = args.distributed_world_size
+
+              def _new_save_checkpoint(*args, **kwargs):
+                  _original_save_checkpoint(*args, **kwargs)
+                  n_cpus = int(ray.cluster_resources()["CPU"])
+                  if n_cpus > original_n_cpus:
+                      raise Exception("New CPUs find (original %d CPUs, now %d CPUs)"
+                                      % (original_n_cpus, n_cpus))
+          else:
+              original_n_gpus = args.distributed_world_size
+
+              def _new_save_checkpoint(*args, **kwargs):
+                  _original_save_checkpoint(*args, **kwargs)
+                  n_gpus = int(ray.cluster_resources().get("GPU", 0))
+                  if n_gpus > original_n_gpus:
+                      raise Exception("New GPUs find (original %d GPUs, now %d GPUs)"
+                                      % (original_n_gpus, n_gpus))
+          fairseq.checkpoint_utils.save_checkpoint = _new_save_checkpoint
+
+          # Call the original main function of fairseq.
+          main(args, init_distributed=(args.distributed_world_size > 1))
+
+      def get_node_ip(self):
+          """Returns the IP address of the current node."""
+          return ray.services.get_node_ip_address()
+
+      def find_free_port(self):
+          """Finds a free port on the current node."""
+          with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+              s.bind(("", 0))
+              s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+              return s.getsockname()[1]
+
+The new main function ``ray_main()`` provides fault-tolerance and keeps the effective batch size the same for different number of GPUs:
+
+.. code-block:: python
+
+  def add_ray_args(parser):
+      """Add ray and fault-tolerance related parser arguments to the parser."""
+      group = parser.add_argument_group('Ray related arguments')
+      # fmt: off
+      group.add_argument('--ray-address', default="auto", type=str,
+                        help='address for ray initialization')
+      group.add_argument('--fix-batch-size', default=None, type=int,
+                        help='fix batch size (max_sentences * update_freq '
+                              '* n_GPUs) to be a fixed input value for different '
+                              'number of GPUs or CPUs')
+      # fmt: on
+      return group
+
+
+  def ray_main():
```

Comment:
can you change this to "run_fault_tolerant_loop"?

Reply:
resolved in the latest comment